lib/wab/impl/shell.rb

```diff
@@ -143,6 +147,42 @@ def start()
         server.start
       end
 
+      # Register a controller for a named type.
+      #
+      # If a request is received for an unregistered type the default controller
+      # will be used. The default controller is registered with a +nil+ key.
+      #
+      # type:: type name
+      # controller:: Controller instance for handling requests for the identified +type+
+      def register_controller(type, controller)
+        controller.shell = self
+        @controllers[type] = controller
+      end
+
+      # Returns the controller associated with the type key found in the
+      # data. If a controller has not be registered under that key the default
+      # controller is returned if there is one.
+      #
+      # data:: data to extract the type from for lookup in the controllers
+      def controller(data)
+        path = data.get(:path)
+        path = path.native if path.is_a?(::WAB::Data)
+        if path.nil? || path.length <= @path_pos
+          content = data.get(:content)
+          return @controllers[content.get(@type_key)] || @controllers[nil] unless content.nil?
+        else
+          return @controllers[path[@path_pos]] || @controllers[nil]
```

Comment:
since we're defining, `:path_controller` below, this line can be replaced with that method instead..
```diff
- return @controllers[path[@path_pos]] || @controllers[nil]
+ path_controller(path)
```